src/java/org/apache/cassandra/metrics/AccordStateCacheMetrics.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.metrics;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.codahale.metrics.Histogram;
+import org.apache.cassandra.cache.CacheSize;
+
+import static org.apache.cassandra.metrics.CassandraMetricsRegistry.Metrics;
+
+public class AccordStateCacheMetrics extends CacheMetrics
+{
+    public final Histogram objectSize;
+
+    /**
+     * Create metrics for given cache.
+     *
+     * @param type  Type of Cache to identify metrics.
+     * @param cache Cache to measure metrics
+     */
+    public AccordStateCacheMetrics(String type, CacheSize cache)
+    {
+        super(type, cache);
+
+        objectSize = Metrics.histogram(factory.createMetricName("ObjectSize"), false);
+    }
+
+    public static class Provider
+    {
+        public static final String ACCORD_STATE_CACHE = "accord-state-cache";
+
+        private final Set<CacheSize> globalCacheSizes;
+        private final AccordStateCacheMetrics globalMetrics;
+
+        private final Map<Class<?>, AccordStateCacheMetrics> instanceMetrics;
+        private final Map<Class<?>, Set<CacheSize>> instanceCacheSizes;
+
+        private final String type;
+
+        public Provider()
+        {
+            this(ACCORD_STATE_CACHE);
+        }
+
+        @VisibleForTesting
+        public Provider(String type)
+        {
+            this.type = type;
+            globalCacheSizes = new CopyOnWriteArraySet<>();
+            globalMetrics = new AccordStateCacheMetrics(type, new CacheSize.Summing(globalCacheSizes));
```

Comment:
I'm a little bit confused about how we're managing the global metrics here. `AccordCommandStores` implements `CacheSize` now, but then we also have this `Summing` component for `globalMetrics` that has to receive and bookkeep the `globalCacheSizes`. Can't we just do something like pass `AccordCommandStores` to the provider on construction, where it can then create a `AccordStateCacheMetrics(type, stores)` (i.e. `stores` is the `AccordCommandStores`) for the globals? That at least gets rid of the `globalCacheSizes` set. (Then `forGlobal()` is simpler and no longer has to be `synchronized` as well.)

Reply:
I'm sorry for that, `AccordCommandStores` implementing `CacheSize` is a leftover and is redundant.

I was thinking about passing the list of command stores explicitly but I chose this solution as it seemed cleaner to me - I mean, there is no mutual dependency between the `Provider` and `AccordCommandStores` and it is easier to unit test.

The synchronization is not a pain point here because it happens once per store initialization.

Reply:
The same thing seems to apply to slow/fast path selection as well...

Reply:
talked to @belliottsmith about that - his point is to collect combined and recovery only stats, so yet another idea...

Reply:
I mean, having `AccordCommandStores` implement `CacheSize` was kind of nice, in any case. Generally, it seems nice to have a situation where metric information can be provided by the classes managing the relevant subjects. Like in this case, `AccordCommandStores` exposing size/capacity/etc. with no knowledge of the fact that we populate metrics w/ that information. It also seemed desirable in the sense that we didn't need the registration element of that in `forGlobal()`. Using `Summing` and registration via the `Provider` doesn't seem incorrect as implemented here...just trying to convince myself it's the simplest/most reasonable approach...